### PR TITLE
fix(Core/QuestNpc): Rework/fix the Stave of Ancients script

### PR DESCRIPTION
**Pull request by vtihomirov**
## Changes Proposed:
- Created a separate npc_stave_of_ancients script, that holds the Demon npc scripts.
- Added a script for each of the npcs. (Artorius, Simone, Precious, Nelson and Franklin)
- Added some sql updates for all of the demon npcs.
    - Added new paths for all of the npc(except Precious which gets its path from creature_formation), generated by the source pathing videos.
    - Added gossip flags.
    - Changed the transformed demon speed since they looked like they are slowed.
    - Added correction to creeping doom move speed and damage modifier application workaround.
    - Removed the loot for all of the npcs when they are in normal human form.
- Added the gossip menus, and gossip options which require the player to have the quest.
- Added in fight despawn mechanic when more than one player or creature is attacking the npc.
- Added in fight despawn for Nelson and Simone when more than 1 player is attacking their minions.
  - Created a blacklist for allowed creatures which will not trigger despawn. Traps and snakes from the snake trap won't/shouldn't trigger the despawn.
- Added party loot prevention mechanic which will clear the loot if the killer doesn't have the quest or if more than 1 unit has attacked the npc.
- Added on spell hit weakness mechanics for the demon npcs.
- Added and corrected Soul Flame to our spell_dbc table.
- Added respawn mechanic to Simone so it will now respawn Precious if Precious is killed and Simone reached home.
- Added NPC despawns to 15 min according to https://classic.wowhead.com/guides/classic-hunter-quest-ancient-petrified-leaf.
- Removed the old winterspring script.

## Issues Addressed:

- https://github.com/azerothcore/azerothcore-wotlk/issues/6966

## SOURCE:

pathing:
    [Artorius](https://youtu.be/UUl8qDWUYnU)
    [Simone](https://youtu.be/wi2Krnrx25I)
    [Franklin](https://youtu.be/vUIFA5JAeNc)
    [Nelson](https://youtu.be/C1NdozH80Yw)

fight mechanics:
    https://www.youtube.com/watch?v=adz-VM7wzo4
    https://www.youtube.com/watch?v=CDXJgNWm0nA
    https://www.youtube.com/watch?v=OCUzUP97UWQ
    https://www.youtube.com/watch?v=whvbpIy36gs
    https://www.youtube.com/watch?v=9fR6KhxpGQE
## Tests Performed:

- Tested without the changes when you go to any of the creatures they don't path and they are not interactable. They instantly transform to demons.
https://youtu.be/eiK2Y982OIk

- Tested with the changes. The creatures now follow a path and are interactable, also play a small RP event before transform.
https://youtu.be/MxsJU6E2ctY

- Tested in game on linux/docker setup from windows client.

## How to Test the Changes:

1. Check if there are any warnings or errors when building.
2. Start the server. Make a hunter character.
3. Check if the creatures have a pathing route.
4. Check if the creatures pathing is clipping or looks irregular in some way.
5. Check if creature have a gossip text.
6. Check if a small RP event plays when you click the gossip option.

##### Testing Artorius the Amiable:

7. Check if the creature acts in the same way as the ones in the source videos.
    * `.go creature id 14531`
    * Check if when you add the quest the gossip option appears. Use `.quest add 7636`
    * Does the creature transform visually to a Demon?
    * Does the creature speaks anything?
    * Does the creature follow and attack you(Is it being agressive?)?
    * Does the creature cast Demonic Doom?
    * Is the creature affected by Stinging Trauma when being hit with Serpent Sting?
    * Is an emote present when you hit the creature with Serpent Sting? It should be.
    * Is the creature periodically removing your Serpent Sting from it and glowing red(using Demonic Enrage)?
    * When you reset the fight with feign death or by activating `.gm on`. Does the creature return to the spot it was when you clicked the gossip option.
    * When you kill the creature in demon form is the quest item present? It should be.
    * When you kill the creature in humanoid form is the quest item present? It shouldn't be.
    * When you engage the creature and another player damages it or when the other player cast a buff on you.
    Does the creature eventually stop in place and dissapear? It should.
    (Note: 0 threat buffs won't trigger the dissapear/despawn)
    * When you use your pet to attack the creature does it make the creature dissapear? It should.
    * When the creature is low on health(and in demon form) and another player or creature kills it with big damage abillity, are you able to get the quest item? No loot should be present.
    * When you select the gossip option and you don't fight the creature for the next 15 minutes. Does the creature dissapear/despawn? It should.

##### Testing Simone the Inconspicuous:

8. Check if the creature acts in the same way as the ones in the source videos.
    * `.go creature id 14527`
    * Check if when you add the quest the gossip option appears. Use `.quest add 7636`
    * Does the creature transform visually to a Demon?
    * Does the creature's pet transform visually to a Demon?
    * Does the creature emote anything?
    * Does the creature speaks anything?
    * Does the creature follow and attack you(Is it being agressive?)?
    * Does the creature's pet also follow and attack you(Is it being agressive?)?
    * Does the creature cast Chain Lightning at you?
    * Are you being periodically affected by Tempress Kiss?
    * Is the creature affected by Silence when being hit by Viper Sting?
    * Is an emote present when you hit the creature with Viper Sting? It should be.
    * When you reset the fight with feign death or by activating `.gm on`. Does the creature return to the spot it was when you clicked the gossip option.
    * When you reset the fight with feign death or by activating `.gm on`. Does the creature's pet also return to the spot where it was when you clicked the gossip option.
    * When you kill the creature in demon form is the quest item present? It should be.
    * When you kill the creature in humanoid form is the quest item present? It shouldn't be.
    * When you engage the creature and another player damages it or its pet or when the other player cast a buff on you.
    Does the creature and its pet eventually stop in place and dissapear? They should dissapear.
    (Note: 0 threat buffs won't trigger the dissapear/despawn)
    * When you kill the creature's pet but not the master and you reset the fight. Will the pet respawn? It should.
    * When you kill the creature and reset the fight does the creature's pet go to the master's corpse? It should.
    * When you kill the creature and the pet is guarding its master's corpse will it dissapear when the master's corpse dissapear? It should.
    * When you use your pet to attack the creature or its pet does it make the creature dissapear? It should.
    * When the creature is low on health(and in demon form) and another player or creature kills it with big damage abillity, are you able to get the quest item? No loot should be present.
    * When you select the gossip option and you don't fight the creature for the next 15 minutes. Does the creature and its pet dissapear/despawn? They should dissapear/despawn.

##### Testing Nelson the Nice:

9. Check if the creature acts in the same way as the ones in the source videos.
    * `.go creature id 14536`
    * Check if when you add the quest the gossip option appears. Use `.quest add 7636`
    * Does the creature transform visually to a Demon?
    * Does the creature speaks anything?
    * Does the creature follow and attack you(Is it being agressive?)?
    * Does the creature cast Soul Flame when you first start the battle?
    * When you start the battle and inflict frost trap on the creature, does the creature Soul Flame aura dissapear? It should.
    * When you did the above step and reset the fight does the creature have the aura? And does the creature cast the Soul Flame again when you engage it? It shouldn't.
    * Is the creature affected by Crippling Clip when being hit with Wing Clip? It should.
    * Is an emote present when you hit the creature with Wing Clip? It should be.
    * Does the creature cast creeping doom and fear in demon form? It should.
    * Is the creeping doom movement speed slow? It should be.
    * Is the creeping doom following you instantly when summoned by the creature? It should.
    * When you reset the fight with feign death or by activating `.gm on`. Does the creature return to the spot it was when you clicked the gossip option.
    * When you kill the creature in demon form is the quest item present? It should be.
    * When you kill the creature in humanoid form is the quest item present? It shouldn't be.
    * When you engage the creature and another player damages it or when the other player cast a buff on you.
    Does the creature eventually stop in place, say a line and dissapear? It should.
    (Note: 0 threat buffs won't trigger the dissapear/despawn)
    * When you use your pet to attack the creature does it make the creature dissapear? It should.
    * When you pet or another player or creature kills one of the creeping doom mobs, does the creature say a line and dissapear? It should.
    * When the creature is low on health(and in demon form) and another player or creature kills it with big damage abillity, are you able to get the quest item? No loot should be present.
    * When you select the gossip option and you don't fight the creature for the next 15 minutes. Does the creature dissapear/despawn? It should.

##### Testing Franklin the Friendly:

10. Check if the creature acts in the same way as the ones in the source videos.
    * `.go creature id 14529`
    * Check if when you add the quest the gossip option appears. Use `.quest add 7636`
    * Does the creature transform visually to a Demon?
    * Does the creature speaks anything?
    * Does the creature follow and attack you(Is it being agressive?)?
    * Is the creature affected by Entropic Sting when being hit with Scorpid Sting?
    * Is the creature periodically removing your Scorpid Sting from it and glowing red(using Demonic Enrage)?
    * Is an emote present when the creature uses Demonic Enrage? It should be
    * When you reset the fight with feign death or by activating `.gm on`. Does the creature return to the spot it was when you clicked the gossip option.
    * When you kill the creature in demon form is the quest item present? It should be.
    * When you kill the creature in humanoid form is the quest item present? It shouldn't be.
    * When you engage the creature and another player damages it or when the other player cast a buff on you.
    Does the creature eventually stop in place, say a line and dissapear? It should.
    (Note: 0 threat buffs won't trigger the dissapear/despawn)
    * When you use your pet to attack the creature does it make the creature dissapear? It should.
    * When the creature is low on health(and in demon form) and another player or creature kills it with big damage abillity, are you able to get the quest item? No loot should be present.
    * When you select the gossip option and you don't fight the creature for the next 15 minutes. Does the creature dissapear/despawn? It should.

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by vtihomirov** [comment]
I haven't added "The Cleaner" because I wasn't able to find it in a more reliable source like a video. I think hes only present for the balance of light and shadow quest,
because the Stave of Ancients npcs already have a way to punish exploiting players e.g. in fight despawn, loot prevention and fools plight casting. Of course if someone can provide a reliable video source of "The Cleaner" spawning, I'll be happy to add him as well.

**Comment by vtihomirov** [comment]
I really took my time on making this PR. I noticed the issue a bit back before it was reported, I started working on this as I wanted to answer my own question "Can I get this fixed with C++?". My initial thought was that if I couldn't handle it I would open an issue about it. So a month ago I noticed that the issue was reported by another member and after that I thought that its maybe time to try and finish this. I continued my work on it as I thought that fixing this with SAI would probably not work very well. I think that SAI additions are very good when you need to make a simple script for a creature, but I think that properly scripting these creatures would require a bit more tweaks than what SAI could offer to me.

TLDR: I wanted to see if can fix this as a newbie in C++.

Sorry about the big PR, but I choosed to create it this big instead of few smaller PRs because its seems like a big hassle for me to maintain and update 4 separate PRs when changes are requested and a merge conflict arises.

Big thanks to anyone that takes his time to review and test this PR. :heart: 
Cheers

**Comment by syssneck** [comment]
What about https://github.com/azerothcore/azerothcore-wotlk/pull/7194 ? Now we have two @Asthalor @vtihomirov

**Comment by vtihomirov** [comment]
@syssneck I'm okay to close this in favor of https://github.com/azerothcore/azerothcore-wotlk/pull/7194. I put this here so we can have another option. Just let me know :slightly_smiling_face:

**Comment by Asthalor** [comment]
Im awaiting some testing on my PR, but i must admit that the development on SAI is long and tedious. We can test this one and see if its good! If not my code is saved on this repository.

**Comment by Kitzunu** [comment]
Why were these creature scripts moved outside of the zone script?

**Comment by vtihomirov** [comment]
> Why were these creature scripts moved outside of the zone script?

Thanks for taking the time to look at this @Kitzunu. I added a new script since 3 of the npcs are on Kalimdor, e.g. Un'Goro, Silithus, Winterspring and Franklin is in Eastern Kingdoms. It seemed wrong to me to pollute the winterspring script I thought of them as world creatures and I added them as such. I could be wrong of course.

**Comment by Asthalor** [comment]
TEST

Simone the Inconspicuous

TL:DR : She dont reset when using feight death. She still keeps the demon form. The rest i thinks its good.

Check if when you add the quest the gossip option appears. Use .quest add 7636 GOOD
Does the creature transform visually to a Demon? GOOD
Does the creature's pet transform visually to a Demon? GOOD
Does the creature emote anything? GOOD
Does the creature speaks anything? GOOD
Does the creature follow and attack you(Is it being agressive?)? NO
Does the creature's pet also follow and attack you(Is it being agressive?)? NO
Does the creature cast Chain Lightning at you? GOOD
Are you being periodically affected by Tempress Kiss? GOOD
Is the creature affected by Silence when being hit by Viper Sting? GOOD
Is an emote present when you hit the creature with Viper Sting? It should be. GOOD
When you reset the fight with feign death or by activating .gm on. Does the creature return to the spot it was when you clicked the gossip option. NO
When you reset the fight with feign death or by activating .gm on. Does the creature's pet also return to the spot where it was when you clicked the gossip option. NO
image

When they appeared they didnt aggroed me instantly, i dont know if that blizz or not.

When you kill the creature in demon form is the quest item present? It should be. GOOD
When you kill the creature in humanoid form is the quest item present? It shouldn't be. GOOD
When you engage the creature and another player damages it or its pet or when the other player cast a buff on you.
Does the creature and its pet eventually stop in place and dissapear? They should dissapear. GOOD
(Note: 0 threat buffs won't trigger the dissapear/despawn)
When you kill the creature's pet but not the master and you reset the fight. Will the pet respawn? It should. Reset didnt work
When you kill the creature and reset the fight does the creature's pet go to the master's corpse? It should. Reset didnt work
When you kill the creature and the pet is guarding its master's corpse will it dissapear when the master's corpse dissapear? It should.
When you use your pet to attack the creature or its pet does it make the creature dissapear? It should. GOOD
When the creature is low on health(and in demon form) and another player or creature kills it with big damage abillity, are you able to get the quest item? No loot should be present.
When you select the gossip option and you don't fight the creature for the next 15 minutes. Does the creature and its pet dissapear/despawn? They should dissapear/despawn.

Videos

https://streamable.com/hz4z5m

Despawning with pet

https://streamable.com/k692ie

**Comment by Asthalor** [comment]
Test Nelson the nice

TL:DR: Same as Simone, the only thing to fix is the reseting should transform them on human again? No idea, but they still stay as demons and they stop moving.


Check if when you add the quest the gossip option appears. Use .quest add 7636 GOOD
Does the creature transform visually to a Demon? GOOD
Does the creature speaks anything? YES LOTS OF BUGS
Does the creature follow and attack you(Is it being agressive?)? GOOD
Does the creature cast Soul Flame when you first start the battle? GOOD
When you start the battle and inflict frost trap on the creature, does the creature Soul Flame aura dissapear? It should. GOOD
When you did the above step and reset the fight does the creature have the aura? And does the creature cast the Soul Flame again when you engage it? It shouldn't. GOOD

Video
https://streamable.com/4eys6a

Is the creature affected by Crippling Clip when being hit with Wing Clip? It should. GOOD
image

Is an emote present when you hit the creature with Wing Clip? It should be. GOOD
Does the creature cast creeping doom and fear in demon form? It should. GOOD
Is the creeping doom movement speed slow? It should be. GOOD
Is the creeping doom following you instantly when summoned by the creature? It should. GOOD
When you reset the fight with feign death or by activating .gm on. Does the creature return to the spot it was when you clicked the gossip option. NO

image

When you kill the creature in demon form is the quest item present? It should be. GOOD
When you kill the creature in humanoid form is the quest item present? It shouldn't be. GOOD 
When you engage the creature and another player damages it or when the other player cast a buff on you.
Does the creature eventually stop in place, say a line and dissapear? It should. GOOD
(Note: 0 threat buffs won't trigger the dissapear/despawn)
When you use your pet to attack the creature does it make the creature dissapear? It should. GOOD
When you pet or another player or creature kills one of the creeping doom mobs, does the creature say a line and dissapear? It should. GOOD
When the creature is low on health(and in demon form) and another player or creature kills it with big damage abillity, are you able to get the quest item? No loot should be present.
When you select the gossip option and you don't fight the creature for the next 15 minutes. Does the creature dissapear/despawn? It should. Did not test

**Comment by Asthalor** [comment]
Test Franklin

TLDR: Instead of dissapear when other creature jumps on me he evades. The buff of demonic frenzy is not showing in the portrait. 

Check if when you add the quest the gossip option appears. Use .quest add 7636 GOOD
Does the creature transform visually to a Demon? GOOD
Does the creature speaks anything? GOOD
Does the creature follow and attack you(Is it being agressive?)? Not instanly
Is the creature affected by Entropic Sting when being hit with Scorpid Sting? GOOD
image

Is the creature periodically removing your Scorpid Sting from it and glowing red(using Demonic Enrage)? GOOD (NO BUFF ON CREATURE)

image

Is an emote present when the creature uses Demonic Enrage? It should be GOOD
When you reset the fight with feign death or by activating .gm on. Does the creature return to the spot it was when you clicked the gossip option. YES, BUT ON DEMON FORM

image

When you kill the creature in demon form is the quest item present? It should be. GOOD
image

When you kill the creature in humanoid form is the quest item present? It shouldn't be. GOOD (Should not be possible tho)
image


When you engage the creature and another player damages it or when the other player cast a buff on you.
Does the creature eventually stop in place, say a line and dissapear? It should. NO HE EVADES
(Note: 0 threat buffs won't trigger the dissapear/despawn)
When you use your pet to attack the creature does it make the creature dissapear? It should. NO HE EVADES
When the creature is low on health(and in demon form) and another player or creature kills it with big damage abillity, are you able to get the quest item? No loot should be present. Did not test
When you select the gossip option and you don't fight the creature for the next 15 minutes. Does the creature dissapear/despawn? It should. Did not test

So far pretty good. Some minor fixes but very good!

**Comment by Asthalor** [comment]
Test Artorius

TLDR: When the combat is droped by using feign death he goes back on demon form and wont move.

Video of feign death 

https://streamable.com/phetcn

Check if when you add the quest the gossip option appears. Use .quest add 7636 GOOD
Does the creature transform visually to a Demon? GOOD
Does the creature speaks anything? GOOD
Does the creature follow and attack you(Is it being agressive?)? Yes, but he wait for you to attack him when far
Does the creature cast Demonic Doom? GOOD
Is the creature affected by Stinging Trauma when being hit with Serpent Sting? GOOD
Is an emote present when you hit the creature with Serpent Sting? It should be. GOOD

Video

https://streamable.com/oij4lc

Is the creature periodically removing your Serpent Sting from it and glowing red(using Demonic Enrage)? YES, BUT NO BUFFS ON PORTRAIT

image

When you reset the fight with feign death or by activating .gm on. Does the creature return to the spot it was when you clicked the gossip option. YES, BUT ON DEMON
When you kill the creature in demon form is the quest item present? It should be. GOOD
When you kill the creature in humanoid form is the quest item present? It shouldn't be. GOOD
When you engage the creature and another player damages it or when the other player cast a buff on you.
Does the creature eventually stop in place and dissapear? It should.
(Note: 0 threat buffs won't trigger the dissapear/despawn) Did not test
When you use your pet to attack the creature does it make the creature dissapear? It should. GOOD

Video

https://streamable.com/3bttdg

When the creature is low on health(and in demon form) and another player or creature kills it with big damage abillity, are you able to get the quest item? No loot should be present. Did not test
When you select the gossip option and you don't fight the creature for the next 15 minutes. Does the creature dissapear/despawn? It should. Did not test

**Comment by Asthalor** [comment]
So to put it into a nutshell:

Lots of things are pretty good, there is only the feign death thing wich i would love for anyone else to help with that to see if its blizz or not and some buffs are not being shown on the demon portraits.

Good job!

**Comment by vtihomirov** [comment]
Thanks a ton for testing and providing a video source @Asthalor :+1: .

**Comment by vtihomirov** [comment]
> TLDR: When the combat is droped by using feign death he goes back on demon form and wont move.

I think that the demons should remain for a given time when they transform to demon form and they should retain the demon form till despawn.
I noticed this behavior here: https://youtu.be/adz-VM7wzo4?t=254, if you take a look at the clock when the player dies its 3:36
and when he engages him again its 3:38. To me this proves it. Also this https://youtu.be/9fR6KhxpGQE?t=130.

**Comment by vtihomirov** [comment]
>When you engage the creature and another player damages it or when the other player cast a buff on you.
Does the creature eventually stop in place, say a line and dissapear? It should. **NO HE EVADES**


Does this mean that when you were attacking the creature and another player also started to attack it, it made the creature evade instead of despawn? Because if it is, its a bug.

**Comment by Asthalor** [comment]
> > When you engage the creature and another player damages it or when the other player cast a buff on you.
> > Does the creature eventually stop in place, say a line and dissapear? It should. **NO HE EVADES**
> 
> Does this mean that when you were attacking the creature and another player also started to attack it, it made the creature evade instead of despawn? Because if it is, its a bug. Edit: Or it only happens with buffs?

Not quite. When i was fighting him another creature started attacking me, making franklin evade and make him go back to his spawn. I will try this afternoon test if on a group for the 4 demons and see how it goes.

Because the 2 things that i mentioned were right i think there is not much else to say.
I will provide videos to check that even buffing a person or healing a person while combating make the demons despawn and we are set.

**Comment by FrancescoBorzi** [comment]
TLDR: status? still WIP or ready to be merged?

**Comment by vtihomirov** [comment]
> TLDR: status? still WIP or ready to be merged?

@FrancescoBorzi I think that status of this is still WIP, since some more tests need to be concluded, also I'm not sure if the code was reviewed.

**Comment by Nyeriah** [comment]
Thank you for your contribution! 

I did a quick review, I commented on the most glaring issues. What really concerns me the most is those lines without null-pointer checks on Update(), those can cause crashes if the objectaccessor fails to find a connected player with that guid that was stored.

**Comment by FrancescoBorzi** [comment]
> Thank you for your contribution!
> 
> I did a quick review, I commented on the most glaring issues. What really concerns me the most is that there are lines without null-pointer checks on Update(), those can cause crashes if the objectaccessor fails to find a connected player using the guid that was stored (e.g player disconnected during the event). Those gotta be addressed.

thanks @Nyeriah , we better mark this as WIP then

**Comment by fendwe** [comment]
See also https://github.com/chromiecraft/chromiecraft/issues/3220
